File: services/blog/src/utils/consumer.ts
import amqp from 'amqplib';
import { redisClient } from '../server.js';
import { sql } from './db.js';

interface CacheInvalidationMessage {
    action: string;
    keys: string[];
}

export const startCacheConsumer = async () => {
    try {
        const connection = await amqp.connect({
            protocol: 'amqp',
            hostname: 'localhost',
            port: 5672,
            username: 'admin',
            password: 'admin123',
        });

        const channel = await connection.createChannel(); 
        const queueName = 'cache-invalidation';  
        await channel.assertQueue(queueName, { durable: true });

        console.log('✅ Blog Cache Consumer connected to RabbitMQ, waiting for messages...');
        channel.consume(queueName, async (msg) => {
            if (msg) {
                try{
                    const content = JSON.parse(msg.content.toString()) as CacheInvalidationMessage;
                    console.log(`📥Blog Service Received cache invalidation message:`, content);

                    if(content.action === "invalidateCache" && content.keys.length > 0) {
                        for(const pattern of content.keys) {
                            const keys = await redisClient.keys(pattern);

                            if(keys.length > 0) {
                                await redisClient.del(keys);
                                console.log(`🗑️ Blog service Invalidated cache for pattern: ${pattern}, keys deleted: ${keys.length}`);

                                const searchQuery = "";
                                const category = "";
                                const cacheKey =  `blogs:${searchQuery}:${category}`;

                                const blogs  =  await sql`SELECT * FROM blogs ORDER BY created_at DESC`;
                                await redisClient.set(cacheKey, JSON.stringify(blogs), { EX: 3600 });
                                console.log(`🔄️ Blog Service Refreshed cache for key: ${cacheKey}`);
                            }
                        }
                    }
                    channel.ack(msg);
                } catch (error) {
                    console.error('❌ Error processing cache invalidation in blog service :', error);
                    channel.nack(msg, false, true);
                }
            }
        });
    } catch (error) {   
        console.error('❌ Cache Consumer failed to connect to RabbitMQ:', error);
    }    
};

File: services/blog/internal/rabbitmq/consumer.go
package rabbitmq

import (
	"context"
	"encoding/json"
	"time"

	"go.uber.org/zap"

	"blog/internal/db"
	"blog/internal/logger"
	"blog/internal/redis"
)

// StartCacheConsumer connects to the queue and listens for invalidation events
func StartCacheConsumer() {
	if Channel == nil {
		logger.Logger.Error("RabbitMQ channel is not established, cannot start consumer")
		return
	}

	queueName := "cache-invalidation"
	_, err := Channel.QueueDeclare(
		queueName, // name
		true,      // durable
		false,     // delete when unused
		false,     // exclusive
		false,     // no-wait
		nil,       // arguments
	)
	if err != nil {
		logger.Logger.Error("Failed to declare cache-invalidation queue", zap.Error(err))
		return
	}

	msgs, err := Channel.Consume(
		queueName, // queue
		"",        // consumer
		false,     // auto-ack (explicit ack)
		false,     // exclusive
		false,     // no-local
		false,     // no-wait
		nil,       // args
	)
	if err != nil {
		logger.Logger.Error("Failed to register a consumer", zap.Error(err))
		return
	}

	logger.Logger.Info("Blog Cache Consumer connected to RabbitMQ, waiting for messages...")

	go func() {
		for d := range msgs {
			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
			var msg CacheInvalidationMessage
			err := json.Unmarshal(d.Body, &msg)
			if err != nil {
				logger.Logger.Error("Error parsing cache invalidation message", zap.Error(err))
				// Reject message without requeue if invalid format
				_ = d.Nack(false, false)
				cancel()
				continue
			}

			logger.Logger.Info("Blog Service Received cache invalidation message", zap.Any("message", msg))

			if msg.Action == "invalidateCache" && len(msg.Keys) > 0 {
				repo := db.NewPostgresBlogRepository()
				for _, pattern := range msg.Keys {
					deleted, err := redis.InvalidateKeysByPattern(ctx, pattern)
					if err != nil {
						logger.Logger.Error("Failed to invalidate keys by pattern", zap.String("pattern", pattern), zap.Error(err))
						continue
					}

					if deleted > 0 {
						// Refill default empty search category list key: blogs:::12:0
						cacheKey := "blogs:::12:0"
						blogs, err := repo.GetAllBlogs(ctx, "", "", 12, 0)
						if err != nil {
							logger.Logger.Error("Failed to fetch blogs from database for cache refill", zap.Error(err))
							continue
						}

						blogsJSON, err := json.Marshal(blogs)
						if err != nil {
							logger.Logger.Error("Failed to marshal blogs for cache refill", zap.Error(err))
							continue
						}

						err = redis.Set(ctx, cacheKey, string(blogsJSON), 3600*time.Second)
						if err != nil {
							logger.Logger.Error("Failed to update cache key after invalidation", zap.String("key", cacheKey), zap.Error(err))
							continue
						}

						logger.Logger.Info("Blog Service Refreshed cache for key", zap.String("key", cacheKey))
					}
				}
			}

			_ = d.Ack(false)
			cancel()
		}
	}()
}
